text/0000-collection-transmute.md

```diff
@@ -0,0 +1,152 @@
+
+```rust
+impl<T> Vec<T> {
+    /// Transmute this `Vec` to a different item type
+    ///
+    /// # Safety
+    ///
+    /// Calling this function requires the target item type be compatible with
+    /// `Self::Item` (see [`mem::transmute`]).
+    ///
+    /// # Examples
+    ///
+    /// transmute a `Vec` of 32-bit integers to their byte representations:
+    ///
+    /// ```
+    /// let x = vec![0u32; 5];
+    /// let y = unsafe { x.transmute::<[u8; 4]>() };
+    /// assert_eq!(5, y.len());
+    /// assert_eq!([0, 0, 0, 0], y[0]);
+    /// ```
+    ///
+    /// [`mem::transmute`]: ../../std/mem/fn.transmute.html
+    unsafe fn transmute<I>(self) -> Vec<I> {
+        ..
+    }
+```
+
+This would mean our example above would become:
+
+```rust
+let x = vec![0i32; 2];
+let y = unsafe { x.transmute::<[u8; 4]>() };
+```
+
+The documentation of `mem::transmute` should link to the new methods.
+
+A clippy lint can catch offending calls to `mem::transmute` and suggest using 
+the inherent `transmute` method where applicable.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation for `Vec` copies the above solution. The methods for 
+`VecDeque` and `BinaryHeap` use the `Vec` method on their data.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Adding a new method to `std` increases code size and needs to be maintained. 
+However, this seems to be a minor inconvenience when compared to the safety 
+benefit.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+As explained above, the method is useful, yet non-obvious. There were multiple 
+wrong implementation attempts before that were not chosen for being unsound.
+
+We could do nothing, but this means users wanting to transmute the items of a 
+`Vec` will be left without an obvious solution which will lead to unsound code 
+if they get it wrong.
+
+We could document the correct solution instead of putting it into `std`. This 
+would lead to worse ergonomics.
+
+We could create a trait to hold the `transmute` method. This would allow more 
+generic usage, but might lead to worse ergonomics due to type inference 
+uncertainty.
+
+It would even be possible to provide a default implementation using 
+`mem::transmute`, but having a default implementation that might be unsound for 
+some types is a footgun waiting to happen.
+
+# Prior art
+[prior-art]: #prior-art
+
+`mem::transmute` offers the same functionality for many other types. We have 
+added similar methods to different types where useful, see the various 
+iterator-like methods in `std`. @Shnatsel and @danielhenrymantilla came up 
+with the solution together in a 
+[clippy issue](https://github.com/rust-lang/rust-clippy/issues/4484).
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- are there other types that would benefit from such a method? What about 
+`HashSet`, `HashMap` and `LinkedList`?
```

Comment:
What is the argument against doing this?

Reply:
For `HashMap` and `HashSet`, you shouldn't be able to do this because it would likely break the hash and values would be in the wrong buckets. `HashMap` provides no api to fix this. `LinkedList` is more amenable to this change, as is `VecDeque`.

Reply:
@KrishnaSannasi 

> For HashMap and HashSet, you shouldn't be able to do this because it would likely break the hash and values would be in the wrong buckets.

We define transmute between two types to be semantically equivalent to a bitwise move.

While reading the RFC it was unclear to me whether the semantics of the proposed methods are a bitwise move of the collection types, or of the elements themselves. 

For the types covered by the RFC, even if the semantics are to perform a bitwise move of the element types, we can optimize that and make the operations O(1) in time and space, so there isn't really a difference in practice.

Whether we can do this for HashMap and HashSet would depend on what the semantics are. For example, if the semantics are to perform a bitwise transmute of each element, then we can just implement these transmutes as a `.into_iter().map(|x| transmute(x)).collect()`. OTOH if these should be O(1), then we can't.

Reply:
I would expect these to be O(1), we could add other methods like `HashMap::map` which would try and reuse the allocation, but that is a separate issue.